sros2/sros2/verb/generate_policy.py

```diff
@@ -126,24 +68,68 @@ def main(self, *, args):
 
         with DirectNode(args) as node:
             for node_name in node_names:
-                profile = self.get_profile(policy, node_name)
+                profile = policy.get_or_create_profile(node_name.node, node_name.ns)
+
                 subscribe_topics = get_subscriber_info(node=node, node_name=node_name)
-                if subscribe_topics:
-                    self.add_permission(
-                        profile, 'topic', 'subscribe', 'ALLOW', subscribe_topics, node_name)
+                for topic in subscribe_topics:
+                    _add_subscribe_topic_expression(profile, node_name, topic)
+
                 publish_topics = get_publisher_info(node=node, node_name=node_name)
-                if publish_topics:
-                    self.add_permission(
-                        profile, 'topic', 'publish', 'ALLOW', publish_topics, node_name)
+                for topic in publish_topics:
+                    _add_publish_topic_expression(profile, node_name, topic)
+
                 reply_services = get_service_info(node=node, node_name=node_name)
-                if reply_services:
-                    self.add_permission(
-                        profile, 'service', 'reply', 'ALLOW', reply_services, node_name)
+                for service in reply_services:
+                    _add_reply_service_expression(profile, node_name, service)
+
                 request_services = get_client_info(node=node, node_name=node_name)
-                if request_services:
-                    self.add_permission(
-                        profile, 'service', 'request', 'ALLOW', request_services, node_name)
+                for service in request_services:
+                    _add_request_service_expression(profile, node_name, service)
 
```

Comment:
Perhaps out of scope for this PR, but can we infer actions from rosgraph yet?
We could add them here when ready.

Reply:
I'm uncertain of the status of that work, but I agree it would be great to have. I also agree it should be another PR though, this one is pretty large.